ngmix/tests/test_em.py

```diff
@@ -0,0 +1,208 @@
+import pytest
+import numpy as np
+
+from ngmix.em import fit_em
+from ngmix.priors import srandu
+from ._sims import get_ngauss_obs
+
+FRAC_TOL = 0.001
+
+
+@pytest.mark.parametrize('noise', [0.0, 0.05])
+def test_em_1gauss(noise):
+    """
+    see if we can recover the input with and without noise to high precision
+    even with a bad guess
+
+    Use ngmix to make the image to make sure there are
+    no pixelization effects
+    """
+
+    rng = np.random.RandomState(42587)
+    ngauss = 1
+    data = get_ngauss_obs(rng=rng, ngauss=ngauss, noise=noise)
+
+    obs = data['obs']
+    gm = data['gmix']
+
+    pixel_scale = obs.jacobian.scale
+
+    pars = gm.get_full_pars()
+    counts = pars[0]
+
+    gm_guess = gm.copy()
+    gm_guess._data["p"] += counts/10 * srandu(rng=rng)
+    gm_guess._data["row"] += 4 * pixel_scale * srandu(rng=rng)
+    gm_guess._data["col"] += 4 * pixel_scale * srandu(rng=rng)
+    gm_guess._data["irr"] += 0.5 * pixel_scale**2 * srandu(rng=rng)
+    gm_guess._data["irc"] += 0.5 * pixel_scale**2 * srandu(rng=rng)
+    gm_guess._data["icc"] += 0.5 * pixel_scale**2 * srandu(rng=rng)
+
+    fitter = fit_em(obs=obs, guess=gm_guess)
+
+    fit_gm = fitter.get_gmix()
+    res = fitter.get_result()
+    assert res['flags'] == 0
+
+    fitpars = fit_gm.get_full_pars()
+
+    if noise == 0.0:
+        assert abs(fitpars[0]/pars[0]-1) < FRAC_TOL
+        assert abs(fitpars[1]-pars[1]) < pixel_scale/10
+        assert abs(fitpars[2]-pars[2]) < pixel_scale/10
+        assert abs(fitpars[3]/pars[3]-1) < FRAC_TOL
+        assert abs(fitpars[4]/pars[4]-1) < FRAC_TOL
+        assert abs(fitpars[5]/pars[5]-1) < FRAC_TOL
+
+    # check reconstructed image allowing for noise
+    imfit = fitter.make_image()
+    imtol = 0.001 / pixel_scale**2 + noise*5
+    assert np.all(np.abs(imfit - obs.image) < imtol)
+
+
+@pytest.mark.parametrize('noise', [0.0, 0.05])
+def test_em_2gauss(noise):
+    """
+    see if we can recover the input with and without noise to high precision
+    even with a bad guess
+
+    Use ngmix to make the image to make sure there are
+    no pixelization effects
+    """
+
+    rng = np.random.RandomState(42587)
+    ngauss = 2
+    data = get_ngauss_obs(rng=rng, ngauss=ngauss, noise=noise)
+    obs = data['obs']
+    gm = data['gmix']
+
+    pixel_scale = obs.jacobian.scale
+
+    pars = gm.get_full_pars()
+    counts_1 = pars[0]
+
+    gm_guess = gm.copy()
+    gm_guess._data["p"] += counts_1/10 * srandu(2, rng=rng)
+    gm_guess._data["row"] += 4 * pixel_scale * srandu(2, rng=rng)
+    gm_guess._data["col"] += 4 * pixel_scale * srandu(2, rng=rng)
+    gm_guess._data["irr"] += 0.5 * pixel_scale**2 * srandu(2, rng=rng)
+    gm_guess._data["irc"] += 0.5 * pixel_scale**2 * srandu(2, rng=rng)
+    gm_guess._data["icc"] += 0.5 * pixel_scale**2 * srandu(2, rng=rng)
```

Comment:
Private data access like this is potentially very brittle. Is there a better way to do this?